### PR TITLE
Improved user control over kernel and material coverage check

### LINKED ISSUE #28117: Improved user control over kernel and material coverage check
### Motivation

As per today, the kernel and material coverage check is "all" or "none":
```
[Problem]
  kernel_coverage_check = true|false
  material_coverage_check = true|false
[]
```

But fully switching off the coverage checks hinders debugging of Moose models. It would be nice to select subdomains to be checked or not to be checked.

### Design

As discussed in #28092, it would be nice to have something like this:

```
kernel_coverage_check = true|on|false|off|skip_list|only_list     #(default is "true")
kernel_coverage_block_list = 'list of subdomains' 
```

and 

```
material_coverage_check = true|on|false|off|skip_list|only_list     #(default is "true")
material_coverage_block_list = 'list of subdomains' 
```

With the additions options "skip_list" and "only_list" the user can exclude/include subdomains explicitly. 

### Impact

Only new functionality is added, but the type of `kernel_coverage_check` and `material_coverage_check` is to be changed to be an enum value. Due to fact, that the new enums also contain "true" and "false" (and "on" an "off") and the corresponding behaviour matches the behaviour today, the negative impact should be minimal.

### PR DESCRIPTION
**Pull request by jmeier**
This PR adds more control on the subdomains the kernel and material coverage check is performed as discussed in #28092.

closes #28117

### CODE REVIEW

**Comment by grunerjmeier** [comment]
Dear Moose Community,

This PR proposes an implementaion for the ideas discussed in #28092.

I would be pleased if you would pay particular attention to these details:

- Instead of `kernel_coverage_blocks` the name `kernel_coverage_block_list` is currently used. I'm happy to change that but I liked to mirror the "list" in `only_list` and `skip_list`
- In FEProblemBase.h there are two new enums (`KernelCoverageCheckMode` and `MaterialCoverageCheckMode`) having identical members. This to allow deviating members in the future. If you dont like this I can merge them to a single enum `CoverageCheckMode`
- I had to alter the signature of method `setKernelCoverageCheck`. Subsequently, there are changes in the uses of this method (e.g. modules)

File: framework/include/problems/FEProblemBase.h [hunk 1]
```diff
@@ -153,6 +153,26 @@ class FEProblemBase : public SubProblem, public Restartable
...
+  {
+    TRUE,
+    ON,
+    FALSE,
+    OFF,
+    SKIP_LIST,
+    ONLY_LIST,
+  };
+
+  enum class MaterialCoverageCheckMode
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:166]: I do prefer having just one enum until a need for two is shown

File: framework/include/problems/FEProblemBase.h [hunk 2]
```diff
@@ -153,6 +153,26 @@ class FEProblemBase : public SubProblem, public Restartable
   FEProblemBase(const InputParameters & parameters);
   virtual ~FEProblemBase();
 
+  enum class KernelCoverageCheckMode
+  {
+    TRUE,
+    ON,
+    FALSE,
+    OFF,
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:161]: These four enumerators should be boiled down to two. I would probably pick `TRUE/FALSE` to reflect the user parameter but I don't have a strong preference

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -205,8 +205,18 @@ FEProblemBase::validParams()
...
+      "kernel_coverage_check",
+      kernel_coverage_check_modes,
+      "Controls, if and how a kernel subdomain coverage check is performed. With 'TRUE' or 'ON' "
+      "all subdomains are checked (the default). Setting 'FALSE' or 'OFF' will disable the check "
+      "for all subdomains. To exclude a predefined set of subdomains 'SKIP_LIST' is to be used, "
+      "while the subdomains to skip are to be defined in the parameter "
+      "'kernel_coverage_block_list'. To limit the check to a list of subdomains, 'ONLY_LIST' is to "
+      "be used (again, using the parameter 'kernel_coverage_block_list').");
+  params.addParam<std::vector<SubdomainName>>(
+      "kernel_coverage_block_list", {}, "List of subdomains for kernel coverage check");
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:219]: Mention in the doc string that this can be a list of blocks to check or blocks to skip the check

File: framework/src/problems/FEProblemBase.C [hunk 4]
```diff
@@ -217,9 +227,18 @@ FEProblemBase::validParams()
...
+  MooseEnum material_coverage_check_modes("TRUE ON FALSE OFF SKIP_LIST ONLY_LIST", "TRUE");
+  params.addParam<MooseEnum>(
+      "material_coverage_check",
+      material_coverage_check_modes,
+      "Controls, if and how a material subdomain coverage check is performed. With 'TRUE' or 'ON' "
+      "all subdomains are checked (the default). Setting 'FALSE' or 'OFF' will disable the check "
+      "for all subdomains. To exclude a predefined set of subdomains 'SKIP_LIST' is to be used, "
+      "while the subdomains to skip are to be defined in the parameter "
+      "'kernel_coverage_block_list'. To limit the check to a list of subdomains, 'ONLY_LIST' is to "
+      "be used (again, using the parameter 'kernel_coverage_block_list').");
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:239]: I would make a lambda in this static `validParams` function that takes a string argument that will either be "material" or "kernel" and returns this long doc-string. It's very long and consequently is likely to be updated in the future. It would be good to only have to update in one place in the code.

File: framework/src/problems/FEProblemBase.C [hunk 5]
```diff
@@ -217,9 +227,18 @@ FEProblemBase::validParams()
...
+  params.addParam<std::vector<SubdomainName>>(
+      "material_coverage_block_list", {}, "List of subdomains for material coverage check");
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:241]: As above, mention that this can be an either "include" or "exclude" list

File: framework/src/problems/FEProblemBase.C [hunk 6]
```diff
@@ -7760,9 +7785,39 @@ FEProblemBase::checkProblemIntegrity()
...
+      _kernel_coverage_check != KernelCoverageCheckMode::FALSE &&
+      _kernel_coverage_check != KernelCoverageCheckMode::OFF)
+  {
+    std::set<SubdomainID> blocks;
+    if (_kernel_coverage_check == KernelCoverageCheckMode::TRUE ||
+        _kernel_coverage_check == KernelCoverageCheckMode::ON)
+      blocks = mesh_subdomains;
+    else if (_kernel_coverage_check == KernelCoverageCheckMode::SKIP_LIST)
+    {
+      blocks.insert(mesh_subdomains.begin(), mesh_subdomains.end());
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:7798]:
```suggestion
      blocks = mesh_subdomains;
```

File: framework/src/problems/FEProblemBase.C [hunk 7]
```diff
@@ -7760,9 +7785,39 @@ FEProblemBase::checkProblemIntegrity()
...
+    }
+    else if (_kernel_coverage_check == KernelCoverageCheckMode::ONLY_LIST)
+      for (const auto subdomain_name : _kernel_coverage_blocks)
+      {
+        const auto id = _mesh.getSubdomainID(subdomain_name);
+        if (id == Moose::INVALID_BLOCK_ID)
+          paramError("kernel_coverage_block_list",
+                     "Subdomain \"" + subdomain_name + "\" not found in mesh.");
+        blocks.insert(id);
+      };
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:7816]:
```suggestion
      }
```

File: framework/src/problems/FEProblemBase.C [hunk 8]
```diff
@@ -7792,6 +7848,33 @@ FEProblemBase::checkProblemIntegrity()
...
+      if (_material_coverage_check == MaterialCoverageCheckMode::SKIP_LIST)
+      {
+        for (const auto subdomain_name : _material_coverage_blocks)
+        {
+          const auto id = _mesh.getSubdomainID(subdomain_name);
+          if (id == Moose::INVALID_BLOCK_ID)
+            paramError("material_coverage_block_list",
+                       "Subdomain \"" + subdomain_name + "\" not found in mesh.");
+          local_mesh_subs.erase(id);
+        };
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:7861]:
```suggestion
        }
```

File: framework/src/problems/FEProblemBase.C [hunk 9]
```diff
@@ -7792,6 +7848,33 @@ FEProblemBase::checkProblemIntegrity()
...
+      {
+        std::set<SubdomainID> blocks(local_mesh_subs);
+        for (const auto subdomain_name : _material_coverage_blocks)
+        {
+          const auto id = _mesh.getSubdomainID(subdomain_name);
+          if (id == Moose::INVALID_BLOCK_ID)
+            paramError("material_coverage_block_list",
+                       "Subdomain \"" + subdomain_name + "\" not found in mesh.");
+          blocks.erase(id);
+        };
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:7873]:
```suggestion
        }
```

File: make_install_test/moose_test/tests/testroot
Comment (lindsayad) [inline make_install_test/moose_test/tests/testroot:1]: was this addition intentional? I don't think this change should be coupled with this PR

File: make_install_test/moose_test/tests/version/tests
Comment (lindsayad) [inline make_install_test/moose_test/tests/version/tests:1]: same comment as above

File: modules/optimization/src/actions/OptimizationAction.C [hunk 10]
```diff
@@ -79,7 +79,7 @@ OptimizationAction::act()
 
     // Set the object parameters
     InputParameters & params = action->getObjectParams();
-    params.set<bool>("kernel_coverage_check") = false;
+    params.set<MooseEnum>("kernel_coverage_check") = false;
```
Comment (lindsayad) [inline modules/optimization/src/actions/OptimizationAction.C:82]: I doubt that this will do the correct thing. This will likely implicitly convert to 0 and consequently correspond to the first of your enumerators (which is `"TRUE"` currently)

File: modules/stochastic_tools/src/actions/StochasticToolsAction.C [hunk 11]
```diff
@@ -84,7 +84,7 @@ StochasticToolsAction::act()
             params.set<bool>("solve") = false;
 
           if (!params.isParamSetByUser("kernel_coverage_check"))
-            params.set<bool>("kernel_coverage_check") = false;
+            params.set<MooseEnum>("kernel_coverage_check") = false;
```
Comment (lindsayad) [inline modules/stochastic_tools/src/actions/StochasticToolsAction.C:87]: use explicit enumerator, e.g. `"false"`

File: modules/stochastic_tools/src/actions/StochasticToolsAction.C [hunk 12]
```diff
@@ -106,7 +106,7 @@ StochasticToolsAction::act()
       // Set the object parameters
       InputParameters & params = action->getObjectParams();
       params.set<bool>("solve") = false;
-      params.set<bool>("kernel_coverage_check") = false;
+      params.set<MooseEnum>("kernel_coverage_check") = false;
```
Comment (lindsayad) [inline modules/stochastic_tools/src/actions/StochasticToolsAction.C:109]: explicit enumerator

File: modules/thermal_hydraulics/test/src/actions/TestAction.C [hunk 13]
```diff
@@ -89,7 +89,7 @@ TestAction::act()
       std::shared_ptr<MooseObjectAction> action = std::static_pointer_cast<MooseObjectAction>(
           _action_factory.create(class_name, "fe_problem", action_params));
 
-      action->getObjectParams().set<bool>("kernel_coverage_check") = false;
+      action->getObjectParams().set<MooseEnum>("kernel_coverage_check") = false;
```
Comment (lindsayad) [inline modules/thermal_hydraulics/test/src/actions/TestAction.C:92]: explicit enumerator

File: test/tests/problems/no_kernel_coverage_check/tests [hunk 14]
```diff
@@ -1,12 +1,29 @@
 [Tests]
-  [./test]
+  [./no_coverage_check]
```
Comment (lindsayad) [inline test/tests/problems/no_kernel_coverage_check/tests:2]:
```suggestion
  [no_coverage_check]
```

File: test/tests/problems/no_kernel_coverage_check/tests [hunk 15]
```diff
@@ -1,12 +1,29 @@
...
     input = 'no_coverage_check.i'
     exodiff = 'no_coverage_check_out.e'
-
     requirement = 'The system shall support the ability to disable the check for'
                   ' PDE operators in all or some parts of the domain.'
     issues = '#2291'
     design = 'Problem/index.md'
-  [../]
+  []
+  [./kernel_coverage_only_list]
```
Comment (lindsayad) [inline test/tests/problems/no_kernel_coverage_check/tests:11]:
```suggestion
  [kernel_coverage_only_list]
```

File: test/tests/problems/no_kernel_coverage_check/tests [hunk 16]
```diff
@@ -1,12 +1,29 @@
...
+    type = 'Exodiff'
+    input = 'kernel_coverage_only_list.i'
+    exodiff = 'kernel_coverage_only_list_out.e'
+    requirement = 'The system shall support the ability to disable the check for'
+                  ' active kernels in some parts of the domain.'
+    issues = '#28117'
+    design = 'Problem/index.md'
+  []
+  [./kernel_coverage_skip_list]
```
Comment (lindsayad) [inline test/tests/problems/no_kernel_coverage_check/tests:20]:
```suggestion
  [kernel_coverage_skip_list]
```

File: test/tests/problems/no_material_coverage_check/material_coverage_only_list.i [hunk 17]
```diff
@@ -0,0 +1,67 @@
+[Problem]
+  solve = true # coverage check needs solve to be true
+
+  kernel_coverage_check = false
+
+  material_coverage_check = 'only_list'
+  material_coverage_block_list = 'BaseMesh'
+  []
```
Comment (lindsayad) [inline test/tests/problems/no_material_coverage_check/material_coverage_only_list.i:8]:
```suggestion
[]
```

File: test/tests/problems/no_material_coverage_check/material_coverage_skip_list.i [hunk 18]
```diff
@@ -0,0 +1,67 @@
+[Problem]
+  solve = true # coverage check needs solve to be true
+
+  kernel_coverage_check = false
+
+  material_coverage_check = 'skip_list'
+  material_coverage_block_list = 'empty_subdomain'
+    []
```
Comment (lindsayad) [inline test/tests/problems/no_material_coverage_check/material_coverage_skip_list.i:8]:
```suggestion
[]
```

File: test/tests/problems/no_material_coverage_check/tests [hunk 19]
```diff
@@ -1,12 +1,29 @@
 [Tests]
-  [./test]
+  [./no_material_coverage_check]
```
Comment (lindsayad) [inline test/tests/problems/no_material_coverage_check/tests:2]:
```suggestion
  [no_material_coverage_check]
```

File: test/tests/problems/no_material_coverage_check/tests [hunk 20]
```diff
@@ -1,12 +1,29 @@
...
     input = 'no_material_coverage_check.i'
     exodiff = 'no_material_coverage_check_out.e'
-
     requirement = 'The system shall support the ability to disable the check for'
                   ' active materials in all parts of the domain.'
     issues = '#5306'
     design = 'Problem/index.md'
-  [../]
+  []
+  [./material_coverage_only_list]
```
Comment (lindsayad) [inline test/tests/problems/no_material_coverage_check/tests:11]:
```suggestion
  [material_coverage_only_list]
```

File: test/tests/problems/no_material_coverage_check/tests [hunk 21]
```diff
@@ -1,12 +1,29 @@
...
+    type = 'Exodiff'
+    input = 'material_coverage_only_list.i'
+    exodiff = 'material_coverage_only_list_out.e'
+    requirement = 'The system shall support the ability to disable the check for'
+                  ' active materials in some parts of the domain.'
+    issues = '#28117'
+    design = 'Problem/index.md'
+  []
+  [./material_coverage_skip_list]
```
Comment (lindsayad) [inline test/tests/problems/no_material_coverage_check/tests:20]:
```suggestion
  [material_coverage_skip_list]
```

File: modules/optimization/src/actions/OptimizationAction.C (hunk 10 above)
Comment (lindsayad) [reply modules/optimization/src/actions/OptimizationAction.C:82]: Use a string instead, e.g. `"false"`

**Review by lindsayad — DISMISSED** [review]

File: framework/include/problems/FEProblemBase.h (hunk 2 above)
Comment (grunerjmeier) [reply framework/include/problems/FEProblemBase.h:161]: Please allow me to double-check on that, @lindsayad: With this PR we change the type of two parameters from bool to enum. Moose [input files](https://mooseframework.inl.gov/application_usage/input_syntax.html) allow for bool beside "TRUE" and "FALASE" also "ON" and "OFF". By only picking "TRUE/FALSE", we invalidate a portion of existing Moose input files (the ones using ON/OFF). Also some of the tests currently in use are using "OFF".
Can I remove "ON" and "OFF" and change all tests used by Moose affected?

Comment (lindsayad) [reply framework/include/problems/FEProblemBase.h:161]: you're right. Let's keep it as is

File: modules/optimization/src/actions/OptimizationAction.C (hunk 10 above)
Comment (GiudGiud) [reply modules/optimization/src/actions/OptimizationAction.C:82]: and let s reorganize the enum to be fool proof to these implicit conversions

false
true
off
on

File: test/tests/problems/no_material_coverage_check/tests [hunk 22]
```diff
@@ -1,12 +1,29 @@
...
     issues = '#5306'
     design = 'Problem/index.md'
-  [../]
+  []
+  [./material_coverage_only_list]
+    type = 'Exodiff'
+    input = 'material_coverage_only_list.i'
+    exodiff = 'material_coverage_only_list_out.e'
+    requirement = 'The system shall support the ability to disable the check for'
+                  ' active materials in some parts of the domain.'
```
Comment (GiudGiud) [inline test/tests/problems/no_material_coverage_check/tests:16]:
```suggestion
    requirement = 'The system shall support the ability to enable the check for'
                  ' active materials only in some parts of the domain.'
```

File: make_install_test/moose_test/tests/version/tests
Comment (grunerjmeier) [reply make_install_test/moose_test/tests/version/tests:1]: This was not intentional! I've no idea, why this is part of my PR.

**Comment by grunerjmeier** [comment]
I've still some failing tests here on my side. I'll investigate when I have access to a bigger screen.

**Comment by grunerjmeier** [comment]
I found it. Now all the tests on my machine are green.

File: framework/src/problems/FEProblemBase.C [hunk 23]
```diff
@@ -7760,9 +7796,38 @@ FEProblemBase::checkProblemIntegrity()
...
+      _kernel_coverage_check != CoverageCheckMode::OFF)
+  {
+    std::set<SubdomainID> blocks;
+    if (_kernel_coverage_check == CoverageCheckMode::TRUE ||
+        _kernel_coverage_check == CoverageCheckMode::ON)
+      blocks = mesh_subdomains;
+    else if (_kernel_coverage_check == CoverageCheckMode::SKIP_LIST)
+    {
+      blocks = mesh_subdomains;
+      for (const auto subdomain_name : _kernel_coverage_blocks)
```
Comment (GiudGiud) [inline framework/src/problems/FEProblemBase.C:7809]:
```suggestion
      for (const auto & subdomain_name : _kernel_coverage_blocks)
```

File: framework/src/problems/FEProblemBase.C [hunk 24]
```diff
@@ -7760,9 +7796,38 @@ FEProblemBase::checkProblemIntegrity()
...
+      {
+        const auto id = _mesh.getSubdomainID(subdomain_name);
+        if (id == Moose::INVALID_BLOCK_ID)
+          paramError("kernel_coverage_block_list",
+                     "Subdomain \"" + subdomain_name + "\" not found in mesh.");
+        blocks.erase(id);
+      }
+    }
+    else if (_kernel_coverage_check == CoverageCheckMode::ONLY_LIST)
+      for (const auto subdomain_name : _kernel_coverage_blocks)
```
Comment (GiudGiud) [inline framework/src/problems/FEProblemBase.C:7819]:
```suggestion
      for (const auto & subdomain_name : _kernel_coverage_blocks)
```

File: framework/src/problems/FEProblemBase.C [hunk 25]
```diff
@@ -7792,6 +7858,33 @@ FEProblemBase::checkProblemIntegrity()
         check_material_coverage = true;
       }
 
+      // did the user limit the subdomains to be checked?
+      if (_material_coverage_check == CoverageCheckMode::SKIP_LIST)
+      {
+        for (const auto subdomain_name : _material_coverage_blocks)
```
Comment (GiudGiud) [inline framework/src/problems/FEProblemBase.C:7864]:
```suggestion
        for (const auto & subdomain_name : _material_coverage_blocks)
```

File: framework/src/problems/FEProblemBase.C [hunk 26]
```diff
@@ -7792,6 +7858,33 @@ FEProblemBase::checkProblemIntegrity()
...
+          if (id == Moose::INVALID_BLOCK_ID)
+            paramError("material_coverage_block_list",
+                       "Subdomain \"" + subdomain_name + "\" not found in mesh.");
+          local_mesh_subs.erase(id);
+        }
+      }
+      else if (_material_coverage_check == CoverageCheckMode::ONLY_LIST)
+      {
+        std::set<SubdomainID> blocks(local_mesh_subs);
+        for (const auto subdomain_name : _material_coverage_blocks)
```
Comment (GiudGiud) [inline framework/src/problems/FEProblemBase.C:7876]:
```suggestion
        for (const auto & subdomain_name : _material_coverage_blocks)
```

**Review by GiudGiud — COMMENTED** [review]
the CI is picky on this

**Comment by GiudGiud** [comment]
Looks like we'll need a Griffin patch for this PR to go in.

**Comment by GiudGiud** [comment]
I am having trouble accessing the INL network right now
@grunerjmeier to be able to pass without patching one of the export controlled application, you ll need to add back
```
setKernelCoverageCheck(bool ...)
```
and 
```
setMaterialCoverageCheck(bool ...)
```
it's ok to have two routines with the same name and different parameters. It's just overloading 

Guillaume

File: framework/src/problems/FEProblemBase.C [hunk 27]
```diff
@@ -217,9 +238,18 @@ FEProblemBase::validParams()
...
+      "material_coverage_check",
+      material_coverage_check_modes,
+      coverage_check_description("material", "material_coverage_block_list"));
+  params.addParam<std::vector<SubdomainName>>(
+      "material_coverage_block_list",
+      {},
+      "List of subdomains for material coverage check. The meaning of this list is controlled by "
+      "the "
+      "parameter 'kernel_coverage_check' (whether this is the list of subdomains to be checked, "
+      "not to be checked or not taken into account).");
```
Comment (GiudGiud) [inline framework/src/problems/FEProblemBase.C:252]:
```suggestion
      "List of subdomains for material coverage check. The meaning of this list is controlled by "
      "the parameter 'material_coverage_check' (whether this is the list of subdomains to be checked, "
      "not to be checked or not taken into account).");
```

**Review by GiudGiud — APPROVED** [review]
tiny item.
Should be able to merge Monday though I m traveling

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
The test failure is because two test requirements are worded exactly the same

**Comment by GiudGiud** [comment]
should be fixed now

**Comment by GiudGiud** [comment]
Failures are time steppers in parallel and web server control. Unrelated

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the contribution! This fits well with the new subdomain modification options
